cmd/geranos/s3uploader.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+
+	"k8s.io/klog/v2"
+)
+
+// see cmd/archeio, this matches the layout of GCR's GCS bucket
+const blobKeyPrefix = "/containers/images/"
+
+type s3Uploader struct {
+	svc          *s3.S3
+	uploader     *s3manager.Uploader
+	skipExisting bool
+}
+
+func newS3Uploader() (*s3Uploader, error) {
+	sess := session.Must(session.NewSession())
+	r := &s3Uploader{}
+	r.svc = s3.New(sess)
+	r.uploader = s3manager.NewUploaderWithClient(r.svc)
+	return r, nil
+}
+
+func (s *s3Uploader) CopyToS3(bucket string, layer v1.Layer) error {
+	key, err := keyForLayer(layer)
+	if err != nil {
+		return err
+	}
+	if s.skipExisting {
+		exists, err := s.blobExists(bucket, key)
+		if err != nil {
+			klog.Errorf("failed to check if blob exists: %v", err)
```

Comment:
Nit: probably should not upload here?  Or if there's a common error case here, a comment would be helpful!

Reply:
It's unclear what an expected error would be here (bad API call??), but uploading again should be a harmless (just costs more time) fall through

Reply:
Or reworded: I think failing towards "just upload it anyhow" is reasonable because skipExisting is only intended to speed up synchronizing. Failing to ensure all content is uploaded to mirrors can cost us $$$ in bandwidth.

I don't think we reasonably expect an error here, but instead of just treating errors as "it doesn't exist" we at least also log them.